Add support for Fractional limits and offsets

0xgouda opened the pull request:
<!--- Disable AI PR formatting assistant: true -->
### Changelog category (leave one):
- New Feature

### Changelog entry (a [user-readable short description](https://github.com/ClickHouse/ClickHouse/blob/master/docs/changelog_entry_guidelines.md) of the changes that goes into CHANGELOG.md):
Add support for fractional `LIMIT` and `OFFSET` for selecting a fraction of a table. Closes #81892

### Documentation entry for user-facing changes
- [x] Documentation is written (mandatory for new features)

### Details
Add support for fractional limits and offset to retrieve percentages of a table for new and old analyzers.

WITH TIES is also supported.

Examples:

```sql
# get the median of the table
SELECT * FROM table ORDER BY id LIMIT 1 OFFSET 0.5
```

```sql
# returns first 20% of the result
SELECT * FROM numbers(10) LIMIT 0.2;
   ┌─number─┐
1. │      0 │
2. │      1 │
   └────────┘
```
```sql
# returns 3rd quartile of the result
SELECT * FROM numbers(12) LIMIT 0.25 OFFSET 0.5;
   ┌─number─┐
1. │      6 │
2. │      7 │
3. │      8 │
   └────────┘
```
```sql
# returns first 10% of the result with ties
SELECT * FROM test ORDER BY id LIMIT 0.1 WITH TIES;
   ┌─id─┐
1. │  0 │
2. │  1 │
3. │  1 │
4. │  1 │
   └────┘
```

Closes #81892

0xgouda committed:
Update Query Analyzer to ccept float numbers in the range `0.1-0.9` for limit and offset

- Used `DataTypeDecimal32(2,1)` for type conversion for now because
for some reason it doesn't recognise `DataTypeFloat32()`.

0xgouda committed:
Update the Limit optimization to check if type is `UInt64`.

0xgouda committed:
Update `QueryAnalysisResult` to parse both `UInt64` and Fractional Limit values.

0xgouda committed:
Use fractional limit/offset values from query analysis to create `Fractional[Limit/Offset]Step`

0xgouda committed:
Update `getLimitLengthAndOffset` to return both `UInt64` and `BFloat16` values of limit/offset.

0xgouda committed:
Add `Fractional[Limit/Offset]Step` creation logic.

0xgouda committed:
Create new `FractionalLimitStep` step header file.

0xgouda committed:
Implement `FractionalLimitStep` methods

0xgouda committed:
Create new `FractionalOffsetStep` class.

0xgouda committed:
Implement `FractionalOffsetStep` class methods

0xgouda committed:
Create new `FractionalLimitTransform` processor class

0xgouda committed:
Implement `FractionalLimitTransform` processor

0xgouda committed:
Create new `FractionalOffsetTransform` processors class

0xgouda committed:
Implement `FractionalOffsetTransform` processor.

0xgouda committed:
Register new `FractionalLimitStep` and `FractionalOffsetStep`

0xgouda committed:
Update `getLimitqueryInfo` to adapt to the new `getLimitLengthAndOffset` signature

0xgouda committed:
Update Exception error message

0xgouda committed:
make function name plural

0xgouda committed:
Update Exception error message

0xgouda committed:
Allow using up to 3 digits after the point in Fractional Limit.

0xgouda committed:
Update comment.

0xgouda committed:
serialize `offset` field

0xgouda committed:
Fix spacing

0xgouda committed:
rename Action to "Fractional Offset"

0xgouda committed:
Update Exception message

0xgouda committed:
ensure chunks_cache is not empty before processing

0xgouda committed:
ensure chunks_cache is not empty before processing

0xgouda committed:
If normal limit + fractional offset, add fractional offset before the limit in the pipe

0xgouda committed:
If normal limit + fractional offset, add fractional offset before the limit in the pipe

0xgouda committed:
if fractional offset add it before limit in the pipeline in `PreliminaryLimitStep` as well

0xgouda committed:
Fix `limit_offset` ignore logic

0xgouda committed:
Refactor `addLimitStep` and `addPreliminaryLimitStep`

0xgouda committed:
Rewrite `FractionalLimitTransform` in a cleaner way

0xgouda committed:
fix: pass `limit_offset` to FractionalOffsetTransform

0xgouda committed:
Refactor `executeLimitStep` and `executePreLimit` to be more readable

0xgouda committed:
don't push limit down if there is a fractional offset

0xgouda committed:
Rewrite `FractionalOffsetTransform` in a cleaner way

0xgouda committed:
Fix styling

0xgouda committed:
Fix comment

0xgouda committed:
Pass sort_description to limit and fractionallimit steps

0xgouda committed:
Fix: update partial sort condition

0xgouda committed:
Remove partial_sorting_limit number used for debugging

0xgouda committed:
Update new analyzer to use Float32

0xgouda committed:
Update old analyzer to use Float32

0xgouda committed:
Update Fractional Limit and Fractional Offset processors to use Float32

0xgouda committed:
Add likely intrinsic

0xgouda committed:
Use Float instead of Bfloat in function name

0xgouda committed:
Refactor addOffsetStep()

0xgouda committed:
Add test for fractional limit and fractiona offset

0xgouda committed:
Don\'t remove sorts if there is fractional limit or offset on top

0xgouda committed:
Don't set  to true on pulling

0xgouda committed:
add pushed rows to  as processed rows, again.

0xgouda committed:
Merge branch 'master' into add-fractional-limit-offset

0xgouda committed:
Merge branch 'master' into add-fractional-limit-offset

0xgouda committed:
Use `lim_info.limit_length` in most checks for readability.

0xgouda committed:
Refactor `executeOffset()`

0xgouda committed:
Imrpvoe conditions readability

0xgouda committed:
Update limit clause docs

0xgouda committed:
Update offset clause docs

0xgouda committed:
fix minor git mistakes

0xgouda committed:
Add comment explaining the processor work

0xgouda committed:
Add comment to make addLimitStep()/executeLimitStep() conditions more clear

0xgouda committed:
Don't count cached rows

0xgouda committed:
Run clang-format

0xgouda committed:
Fix compilation error

0xgouda committed:
Reorder Limit Parsing, try Int28 before Float32

0xgouda committed:
Add more testing to fractional limit test

0xgouda committed:
Add test for distributed fractional limit and offset queries

0xgouda committed:
Fix some styling in Planner.cpp

0xgouda committed:
Fix some styling issues in InterpreterSelectQuery.cpp

0xgouda committed:
Fix more style errors

0xgouda committed:
Fix: still use LimitStep if Limit 0

0xgouda committed:
Fix duplicate include

0xgouda committed:
Fix addOffsetStep() condition

0xgouda committed:
Fix executeOffset() conditions

0xgouda committed:
Revert Planner.cpp style changes

0xgouda committed:
fix docs styling

nihalzp commented:
`03681_distributed_fractional_limit_offset`
`03653_fractional_limit_offset`

Tests are failing for "Stateless tests (amd_binary, ParallelReplicas, s3 storage, parallel)"

```txt
2025-10-18 15:41:59 Reason: result differs with reference:  
2025-10-18 15:41:59 --- /home/ubuntu/actions-runner/_work/ClickHouse/ClickHouse/tests/queries/0_stateless/03681_distributed_fractional_limit_offset.reference	2025-10-18 15:34:29.631061976 -0300
2025-10-18 15:41:59 +++ /home/ubuntu/actions-runner/_work/ClickHouse/ClickHouse/tests/queries/0_stateless/03681_distributed_fractional_limit_offset.stdout	2025-10-18 15:41:59.128608774 -0300
2025-10-18 15:41:59 @@ -1,12 +1,6 @@
2025-10-18 15:41:59  Old Analyzer:
2025-10-18 15:41:59  0
2025-10-18 15:41:59  99
2025-10-18 15:41:59 -90
2025-10-18 15:41:59 -90
2025-10-18 15:41:59 -90
2025-10-18 15:41:59  New Analyzer:
2025-10-18 15:41:59  0
2025-10-18 15:41:59  99
2025-10-18 15:41:59 -90
2025-10-18 15:41:59 -90
2025-10-18 15:41:59 -90
2025-10-18 15:41:59 
2025-10-18 15:41:59 
```

0xgouda committed:
a small update, to get tests re-runned

0xgouda commented:
@nihalzp 
I thought they were unrelated, that's weird, I am on a travel, will debug this when I am back, so please tolerate a few days delay

Also, is there any planner/optimizer logic/step regarding distributed queries that I may have missed?, Because I can't reason why it fails for the distributed tests only

0xgouda commented:
Well, think I know why it fails,

For a distributed `limit m` and/or `limit -m`, the planner spawns a processor for each address, one of them additionally aggregates the results from the other via its input. For a limit or negative limit, this is not a problem, as the processor knows exactly how much to skip (a fixed number of limit/offset). So it only affects its own chunk.

But for a fractional limit at the end of each pull stage in every processor, we calculate `rows_cnt * fraction`, so this extra processor will reapply the fraction processing on chunks already processed by other processors at its input. resulting in incorrect results.

what i can think of is that we should separate these processors and an extra union layer without one processor havingg its output piped into another fractional limit/offset processor?

0xgouda committed:
Through exception on using negative limit with fractional offset or vise verca

0xgouda committed:
fix styling

nihalzp commented:
There is no rush. Take your time :)

I am not immediately sure. I feel the issue could be in `Planner.cpp` and `InterpreterSelectQuery.cpp`. One strategy could be that since `FractionalLimitTransform` is similar to `LimitTransform`, you could double check whether the usage of both transform in those two files are same are not (obviously, you will need to disable many optimisations for `FractionalLimitTransform` because we do not know beforehand the actual offset and limit). The rational comes from the fact that if someone knows total number of rows of table beforehand, they can produce `FractionalLimitTransform` functionality in terms of `LimitTransform`.

0xgouda committed:
Fix: if `do_not_skip_offset` set, add fractional_offset to fractional_limit

0xgouda committed:
Don't use prelimit if there is fractional limit/offset

0xgouda committed:
Don't do partial sort if fractional_limit specified

0xgouda committed:
fix chunk ignore condition

0xgouda committed:
Improve the distributed test

0xgouda committed:
Update limit/offset docs

0xgouda commented:
ping @nihalzp

0xgouda committed:
Add `is_input_port_finished` to `PortsData`

- additional check to avoid counting a finished port more than once.

0xgouda committed:
Fix style error

0xgouda committed:
Make some comments and variable names more readable

0xgouda committed:
Throw exception on using fractional/negative limit/offset in limit by expression

0xgouda committed:
Merge branch 'master' into add-fractional-limit-offset

0xgouda committed:
remove 'LIMIT -1 BY' from test, its now an exception

0xgouda committed:
Merge branch 'master' into add-fractional-limit-offset

nihalzp commented on docs/en/sql-reference/statements/select/limit.md:
```diff
@@ -20,6 +20,19 @@ Negative limits are also supported:
+    • `LIMIT 0.5 OFFSET 0.25` selects 3rd quartile of the result
+
+> **Note**
+> • Fractions must be Float32 numbers less than 1 and greater than zero
```
Why not `Float64`? Then user will be able to provide more precise the limit value.

0xgouda replied:
Done, but now with float64, `old_analyzer` has an imprecision of +1 or -1. I debugged it to ensure it's not a problem anywhere else, but I ensured it's just a rounding problem. This imprecision doesn't appear in `new_analyzer`, probably due to different handling of the constant node.

Example: `select * from numbers(10) limit 0.1` here, the `old_analyzer` will return 2 rows (an +1 imprecision), but as the row numbers grow, it will get identical to `new_analyzer`, which is the more precise one for some reason.

So in `03653_fractional_limit_offset.{sql,reference}` you can see that the `old_analyzer` results differ by +1,-1 from `new_analyzer`

0xgouda replied:
Done.

nihalzp commented on src/Analyzer/Resolve/QueryAnalyzer.cpp:
```diff
@@ -665,10 +668,24 @@ void QueryAnalyzer::convertLimitOffsetExpression(QueryTreeNodePtr & expression_n
     throw Exception(ErrorCodes::INVALID_LIMIT_EXPRESSION,
-        "{} numeric constant expression is not representable as UInt64 or Int64",
-        expression_description);
+        "The value {} of {} expression is not representable as UInt64 or Int64 or Decimal32(8, 7) range [0.1 to 0.9]",
```
Use `(0, 1)` since `LIMIT 0.00001` is also valid.

nihalzp commented on src/Interpreters/InterpreterSelectQuery.cpp:
```diff
@@ -3187,18 +3211,22 @@ void InterpreterSelectQuery::executeDistinct(QueryPlan & query_plan, bool before
         LimitInfo lim_info = getLimitLengthAndOffset(query, context);
 
+        if (lim_info.fractional_offset > 0 || lim_info.fractional_limit > 0)
+            // Preliminary Limits mustn't be added if there is a fractional limit/offset
```
Could you add bit more reasoning in the comment on why prelim should not be applied? It will be useful for future developers.

nihalzp commented on src/Interpreters/InterpreterSelectQuery.cpp:
```diff
@@ -3335,7 +3370,19 @@ void InterpreterSelectQuery::executeLimit(QueryPlan & query_plan)
 
-        if (!lim_info.is_limit_length_negative && !lim_info.is_limit_offset_negative) [[likely]]
+        if (lim_info.is_limit_length_negative && lim_info.fractional_offset > 0)
+            throw Exception(ErrorCodes::BAD_ARGUMENTS, "Negative Limits can't have a fractional Offset");
```
Let's use `NOT_IMPLEMENTED` instead of `BAD_ARGUMENTS`. In the future, we might want to support it as well.
```suggestion
            throw Exception(ErrorCodes::NOT_IMPLEMENTED, "Negative LIMIT is not supported with fractional OFFSET");
```

nihalzp commented on src/Interpreters/InterpreterSelectQuery.h:
```diff
@@ -14,6 +14,7 @@
 #include <Storages/SelectQueryInfo.h>
 #include <Storages/TableLockHolder.h>
 #include <QueryPipeline/Pipe.h>
+#include <base/BFloat16.h>
```
```suggestion
```

nihalzp commented on src/Planner/Planner.cpp:
```diff
@@ -989,14 +1017,27 @@ void addWithFillStepIfNeeded(QueryPlan & query_plan,
+            ErrorCodes::INVALID_LIMIT_EXPRESSION,
+            "The value {} of LIMIT in LIMIT BY expression is not representable as UInt64",
+            applyVisitor(FieldVisitorToString(), limit_by_limit_field));
+
```
We should not have entered the function for negative or fractional limit. This function should only be entered for nonnegative integral limit. So, this check and exception seems redundant.

Please see the comment above around line 455.

nihalzp commented on src/Planner/Planner.cpp:
```diff
@@ -1279,10 +1326,24 @@ void addLimitStep(QueryPlan & query_plan,
 
-    if (!is_limit_length_negative && !is_limit_offset_negative) [[likely]]
+    if (is_limit_length_negative && fractional_offset > 0)
+        throw Exception(ErrorCodes::BAD_ARGUMENTS, "Negative Limits can't have a fractional Offset");
```
```suggestion
        throw Exception(ErrorCodes::NOT_IMPLEMENTED, "Negative LIMIT is not supported with fractional OFFSET");
```

Maybe in the future, we will...

nihalzp commented on src/Planner/Planner.cpp:
```diff
@@ -1279,10 +1326,24 @@ void addLimitStep(QueryPlan & query_plan,
+        throw Exception(ErrorCodes::BAD_ARGUMENTS, "Fractional Limits can't have a negative Offset");
+
+    // only one of limit_length or fractional_limit will have a value not both
+    // only one of limit_offset or fractional_offset will have a value not both
```
Same comment as the old analyzer.

nihalzp commented on src/Planner/Planner.cpp:
```diff
@@ -1338,18 +1426,28 @@ void addExtremesStepIfNeeded(QueryPlan & query_plan, const PlannerContextPtr & p
-    /// If there is not a LIMIT but an offset
-    if (!query_analysis_result.limit_length && query_analysis_result.limit_offset)
+    /// If there is a LIMIT or there is no OFFSET
+    if (!query_analysis_result.limit_length && query_analysis_result.fractional_limit == 0
```
```suggestion
    if ((!query_analysis_result.limit_length && query_analysis_result.fractional_limit == 0)
```

nihalzp commented on src/Planner/PlannerJoinTree.cpp:
```diff
@@ -79,6 +80,8 @@
 #include <Planner/CollectSets.h>
 #include <Planner/CollectTableExpressionData.h>
+#include <base/BFloat16.h>
+#include <base/types.h>
```
```suggestion
```

nihalzp commented on src/Processors/QueryPlan/ReadFromSystemNumbersStep.cpp:
```diff
@@ -407,7 +408,7 @@ std::optional<size_t> getLimitFromQueryInfo(const SelectQueryInfo & query_info,
     const auto lim_info = InterpreterSelectQuery::getLimitLengthAndOffset(query_info.query->as<ASTSelectQuery &>(), context);
 
-    if (lim_info.is_limit_length_negative)
+    if (lim_info.is_limit_length_negative || lim_info.fractional_offset > 0)
```
I am curious. Why do not we skip it for `lim_info.fractional_limit > 0`?

nihalzp commented on tests/queries/0_stateless/03570_limit_by_all.sql:
```diff
@@ -191,11 +191,6 @@ FROM test_limit_by_all
-SELECT id, category, value
-FROM test_limit_by_all
-ORDER BY id, category, value
-LIMIT -1 BY id;
```
Thanks for catching and fixing it!
In most cases, we should not remove a test if the output is fixed. Let's keep the test and add -- { serverError ... }

Additionally, maybe we can duplicate the test for `old_analyzer` as well?

0xgouda replied:
reverted with `serverError NOT_IMPLEMENTED`.

`old_analyzer` doesn't support `LIMIT BY ALL`, should we explicitly list all the columns for it?

nihalzp replied:
Okay, that would be reasonable.

0xgouda replied:
It turned out that this file specifically focuses on the `LIMIT BY ALL` clause, so I instead added extended coverage for `LIMIT BY` for `old_analyzer` in `03571_limit_by_all_old_planner.sql`

nihalzp reviewed (changes requested):
The functionality seems to be working. Especially, good job on `FractionalLimitTransform` which was tricky!

I have left some comments. They apply to other places as well, but in most cases I added the note only once. Let me know if you have any questions.

Additionally, could we have more tests on `WITH TIES`?

0xgouda commented on src/Interpreters/InterpreterSelectQuery.cpp:
```diff
@@ -206,7 +217,7 @@ namespace ServerSetting
 }
 
-static std::pair<UInt64, bool> getLimitOffsetAbsAndSign(const ASTPtr & node, const ContextPtr & context, const std::string & expr);
+static std::tuple<UInt64, bool, Float32> getLimitOffsetAbsAndSignAndFraction(const ASTPtr & node, const ContextPtr & context, const std::string & expr);
```
> Let's do return type std::tuple<UInt64, Float32, bool>. This way it will make more sense in the future if we also support negative fractional limit.

Float32 can hold Negative values, but yeah, that will look cleaner anyway.

nihalzp commented on src/Processors/QueryPlan/FractionalLimitStep.cpp:
```diff
@@ -0,0 +1,117 @@
+    writeIntBinary(flags, ctx.out);
+
+    writeFloatText(limit_fraction, ctx.out);
+    writeFloatText(offset_fraction, ctx.out);
```
If there are the multiple servers, then for distributed queries, they are used for shipping query plan to other servers. If two servers are running, first server might call `serialize()` to encode its configuration of a step (e.g. `FractionalLimitStep`); then the other server calls deserialize() to reconstruct the step and execute it.

0xgouda committed:
Improve Exception Message in `convertLimitOffsetExpression()`

0xgouda committed:
Rename `getLimitOffsetAbsAndSignAndFraction()` to `getLimitOffsetValue()`

0xgouda committed:
Throw Exceptions on invalid lim_info status

- both `limit_length` and `fractional_limit` have non-zero value
- ...

0xgouda committed:
remove unused headers

0xgouda committed:
Explicitly return 0 if `fractional_limit > 0`

0xgouda committed:
Serialize Float32 as binary instead of text.

0xgouda committed:
Fix typo

0xgouda committed:
Rename `getLimitAbsAndSignAndFraction()` to `getLimitOffsetValue()`

0xgouda committed:
Throw logical exceptions on invalid limit values state.

- both `limit_length` and `fractional_limit` have non-zero value.
- ...

0xgouda committed:
Improve fractional limit docs clarity

0xgouda committed:
Improve fractional offset docs clarity

0xgouda commented on docs/en/sql-reference/statements/select/limit.md:
```diff
@@ -20,6 +20,19 @@ Negative limits are also supported:
+`LIMIT 0.m` selects the first (m * 10)% percent of rows from the result.
+
+`LIMIT 0.m OFFSET 0.n` selects the (m * 10)% percent of rows from the result, after skipping the first (n * 10)% percent of rows.
+
```
Tried to make it more clear, please re-review.

0xgouda committed:
Fix style error

0xgouda committed:
Check `hasLimitByOffset()` before analyzing limit by offset

0xgouda committed:
Revert test removal and add `serverError NOT_IMPLEMENTED`.

0xgouda committed:
Move with ties testing for fractional limit to a separate test file.

0xgouda committed:
Use `Float64` in fractional limit/offset instead of `Float32`

0xgouda committed:
Improve Clarity

- remove duplicate `percent`
- end sentences with a period.

0xgouda committed:
Add extended coverage for `LIMIT BY` in `old_analyzer`

nihalzp commented on docs/en/sql-reference/statements/select/limit.md:
```diff
@@ -14,12 +14,30 @@ doc_type: 'reference'
 
+Selecting a fraction of the result is also supported:
+
+`LIMIT m` - If 0 < m < 1 then the first (m * 100)% percent from the result is returned.
```
```suggestion
`LIMIT m` - if 0 < m < 1 then the first m * 100% of rows are returned.
```

nihalzp commented on docs/en/sql-reference/statements/select/limit.md:
```diff
@@ -14,12 +14,30 @@ doc_type: 'reference'
+
+Examples:
+    • `LIMIT 0.1` - selects the first 10% of the result.
+    • `LIMIT 0.5 OFFSET 1` - selects the median row.
```
```suggestion
    • `LIMIT 1 OFFSET 0.5` - selects the median row.
```

nihalzp commented on docs/en/sql-reference/statements/select/limit.md:
```diff
@@ -14,12 +14,30 @@ doc_type: 'reference'
 
 `LIMIT -m OFFSET -n` selects the last `m` rows after skipping the last `n` rows. The `LIMIT -n, -m` syntax is equivalent.
 
+Selecting a fraction of the result is also supported:
```
```suggestion
Moreover, selecting a fraction of the result is also supported:
```

nihalzp commented on src/Planner/Planner.cpp:
```diff
@@ -348,26 +360,37 @@ void extendQueryContextAndStoragesLifetime(QueryPlan & query_plan, const Planner
-    return {magnitude, true};
+    throw Exception(
+        ErrorCodes::INVALID_LIMIT_EXPRESSION,
+        "The value {} of LIMIT/OFFSET expression is not representable as UInt64 or Int64 or Float64 range [0 - 1)",
```
Same comment as the old analyzer.

nihalzp commented on src/Planner/Planner.cpp:
```diff
@@ -989,13 +1032,11 @@ void addWithFillStepIfNeeded(QueryPlan & query_plan,
 void addLimitByStep(
     QueryPlan & query_plan, const LimitByAnalysisResult & limit_by_analysis_result, const QueryNode & query_node, bool do_not_skip_offset)
 {
-    /// Constness of LIMIT BY limit is validated during query analysis stage
```
Let's keep the old comments.

nihalzp reviewed (changes requested):
It looks really good!

I added few comments that I missed last time. They are mostly small changes to improve user-facing messages.

I think after the fixes, it should be ready to merge :)

0xgouda committed:
Improve typing clarity in `limit/offset` docs.

0xgouda committed:
Use `(0, 1)` instead of `[0, 1)` and `0.1 - 0.9` in Fractional{Limit,Offset}

0xgouda commented on docs/en/sql-reference/statements/select/limit.md:
```diff
@@ -14,12 +14,30 @@ doc_type: 'reference'
+    • `LIMIT 0.25 OFFSET 0.5` - selects 3rd quartile of the result.
+
+> **Note**
+> • The fraction must be a Float64 number less than 1 and greater than zero.
```
hyperlink to what?

nihalzp replied:
Similar to:

```txt
- `value` — Input values, must be [Integer](../../data-types/int-uint.md) or [Float64](../../data-types/float.md) type.
```

0xgouda committed:
Through exceptions on invalid `lim_info` state for `OFFSET` executors.

0xgouda committed:
Fix style errors

nihalzp commented on src/Interpreters/InterpreterSelectQuery.cpp:
```diff
@@ -2126,9 +2145,10 @@ void InterpreterSelectQuery::executeImpl(QueryPlan & query_plan, std::optional<P
+            if (try_apply_prelimit)
             {
-                executePreLimit(query_plan, /* do_not_skip_offset= */!apply_offset);
+                applied_prelimit = executePreLimit(query_plan, /* do_not_skip_offset= */!apply_offset);
```
Performance is not a issue because your query will likely to be processing millions or billions rows; in that case, spending few more CPU cycles on this will be just unmeasurable noise.

The issue with current flow is that if someone is reading the code, they now need to go inside `executePreLimit` to understand it's logic because it's return value now influences decisions of the parent function down the line. On the other hand, your version in `planner.cpp` is much easier to understand; future programmars will not need to know what's happening inside `executePreLimit` and treat it like a blackbox.

nihalzp reviewed (commented):
I noticed that during transforms, the algorithm first caches all chunks; afterwards, when all pulling is done, only then it starts processing the chunks. The issue is that it will have slow and high memory usage and would not work if the entire table size is bigger than available memory.

For example:
```sql
SELECT *
FROM numbers(1e9)
OFFSET 0.99999999
```
```txt
10 rows in set. Elapsed: 3.646 sec. Processed 990.88 million rows, 7.93 GB (271.76 million rows/s., 2.17 GB/s.)
Peak memory usage: 7.45 GiB.
```
Here we are caching the entire table of 1e9 rows and that's why memory usage is high. If we used a bigger table whose size is bigger than available RAM, the query would fail.

But if we keep evicting front chunks which will never be part of the final result, we can get very low memory footprint:
```sql
SELECT *
FROM numbers(1e9)
LIMIT -10
```

```txt
10 rows in set. Elapsed: 0.328 sec. Processed 919.59 million rows, 7.36 GB (2.80 billion rows/s., 22.41 GB/s.)
Peak memory usage: 17.27 KiB.
```

Offset version:
```sql
SELECT *
FROM numbers(1e9)
OFFSET 1e9 - 10
```

```txt
10 rows in set. Elapsed: 0.320 sec. Processed 943.98 million rows, 7.55 GB (2.95 billion rows/s., 23.61 GB/s.)
Peak memory usage: 9.23 KiB.
```

For context, we can do some inference about already cache chunks and either discard or push them to free up memory. For example, suppose we have a query with OFFSET 0.5. At one point in time, before we have pulled all the data, we have cached chunks which have a total of 100 rows. Now, even if we do not yet know the total number of rows in the table, we can safely infer that the first 50 rows can be discarded, and we can discard cached complete chunks within the first 50 rows.

Similar for LIMIT, instead of discard, we would push to output ports.

This will require some extra bookkeeping but it can make the feature really high quality and blazingly fast!

0xgouda committed:
Add hyperlinks to Float64 type page

0xgouda committed:
Add comment to explain why we shouldn't use prelimits if there is a fractional limit/offset.

0xgouda committed:
Don't enter `executePreLimit()` if there is a fractional limit/offset

- add additinoal `getLimimtLengthAndOffset()` parsing step to ensure
fractional limit and offset are zero before adding a prelimit.

0xgouda commented:
> I think optimizing for `FractionalOffsetTransform` should be simple change

I am not sure if optimizing `FractionalOffsetTransform` is going to be that simple, because we will remove the push phase entirely and in the pull phase every couple of chunks apply the offset on them as you proposed and at the very end push all remaining chunks, but the point is (i am not very sure) there might be scenarios where we still want to offset in the remaining chunks (the ones to be pushed at the end) then maintaining the offset integer till that final phase is going to be tricky i guess.

I would prefer working on these + some more optimizations like limit pushdown, etc. in a separate PR This one already got too big.

nihalzp commented:
> I was planning to implement WITH TIES for the negative limit processor next, but I think I should go for the optimizations you proposed first.

Anything that interests you would be very much welcome :)

nihalzp reviewed (approved):
Looks good to me!

Just few small doc additions I thought would be better if we are more precise.

nihalzp commented on docs/en/sql-reference/statements/select/limit.md:
```diff
@@ -14,12 +14,30 @@ doc_type: 'reference'
 
+Moreover, selecting a fraction of the result is also supported:
+
+`LIMIT m` - if 0 < m < 1, then the first m * 100% of rows are returned.
```
You are right. My suggestion is not the actual behavior. Let's ignore it.

nihalzp replied:
```suggestion
`LIMIT m` - if 0 < m < 1, then the first `ceil(m × total_rows)` rows are returned.
```

What do you think of this?

nihalzp replied:
Small grammar improvement:

`If a fractional number of rows resulted from the calculation, Its rounded up to the next whole number.`

to

`If a fractional number of rows results from the calculation, it is rounded up to the next whole number.`

0xgouda committed:
Add note about rounding up the fractional limit/offset calculation result.

nihalzp committed:
Merge remote-tracking branch 'upstream/master' into add-fractional-limit-offset

nihalzp committed:
Rerun CI

nihalzp commented:
Failed integration tests:
* https://github.com/ClickHouse/ClickHouse/issues/89758
